### PR TITLE
Addon non-Halide code blocks for linkage.

### PR DESCRIPTION
**Pull request by zvookin**
(Opening review to get comments. This depends on submodules_lowering and may not be fully complete yet.)

This provides a mechanism to attach blocks of external code to a Module. These will be added to or linked in with all code generated for a certain target. This can be used to bind hand written assembly or externally compiled helper functions in with Halide generated object code. For top-level compilations via LLVM, this is somewhat of a minor convenience at the code could just be linked in when the final executable is compiled. But for object code that is generated for an offload target, such as Hexagon, there is no separate linkage mechanism so it must be done by Halide.

Ultimate goal is to support adding source code to C/JavaScript/OpenCL/etc. and other forms of extension.

It is easy to encapsulate both externally called functionality and the code that backs it up in a Generator.

### CODE REVIEW

**Comment by zvookin** [comment]
In particular, it needs tests.

File: src/ExternalCode.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#include "Target.h"
+
+namespace Halide {
+
+class ExternalCode {
+private:
+    enum Kind {
+        LLVMBitcode,
+        DeviceCode,
+        CXXSource,
```
Comment (abadams) [inline src/ExternalCode.h]: Elsewhere we use CPlusPlus

File: src/ExternalCode.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    static ExternalCode device_code_wrapper(DeviceAPI device_api, const std::vector<uint8_t> &code, const std::string &name) {
+        return ExternalCode(DeviceCode, Target(), device_api, code, name);
+    }
+
+    static ExternalCode cxx_code_wrapper(const std::vector<uint8_t> &code, const std::string &name) {
+        return ExternalCode(CXXSource, Target(), DeviceAPI::None, code, name);
+    }
+
+    bool is_for_cpu_target(const Target &host) const {
+      // ArchUnknown can be used to insert generitc llvm bitcode.
```
Comment (abadams) [inline src/ExternalCode.h]: s/generitc/generic

File: src/ExternalCode.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      // has access to the feature flags in effect and can select code qppropriately.
+      return kind == LLVMBitcode &&
+        (llvm_target.arch == Target::ArchUnknown || llvm_target.arch == host.arch) &&
+        (llvm_target.os == Target::OSUnknown || llvm_target.os == host.os) &&
+        (llvm_target.bits == host.bits);
+    }
+    bool is_for_device_api(DeviceAPI current_device) const {
+        return kind == DeviceCode && device_code_kind == current_device;
+    }
+    bool is_cxx_source() const { return kind == CXXSource; }
```
Comment (abadams) [inline src/ExternalCode.h]: c_plus_plus?

File: src/Generator.h [hunk 4]
```diff
@@ -2185,6 +2186,11 @@ class GeneratorContext {
 public:
     virtual ~GeneratorContext() {};
     virtual Target get_target() const = 0;
+
+    // TODO: naming, exact contents of map
```
Comment (abadams) [inline src/Generator.h]: TODO

File: src/ExternalCode.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    // Used for debugging and naming the module to llvm.
+    std::string nametag;
+    
+    ExternalCode(Kind kind, const Target &llvm_target, DeviceAPI device_api, const std::vector<uint8_t> &code, const std::string &name)
+        : kind(kind), llvm_target(llvm_target), device_code_kind(device_api), code(code), nametag(name) {
+    }
+
+public:
+
+    // TODO: doc strings
```
Comment (abadams) [inline src/ExternalCode.h]: TODO here as well. No opinion on using the rval reference.

File: src/Module.cpp [hunk 6]
```diff
@@ -266,8 +275,27 @@ Module Module::resolve_submodules() const {
     for (const auto &buf : buffers()) {
         lowered_module.append(buf);
     }
+    for (const auto &ec : external_code()) {
+        lowered_module.append(ec);
+    }
     for (const auto &m : submodules()) {
         Module copy(m.resolve_submodules());
+
+        // Propgate external code blocks.
```
Comment (abadams) [inline src/Module.cpp]: Propgate -> Propagate

File: test/correctness/external_code.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    module.append(external_code);
+
+    auto forty_two = module.get_function_by_name("forty_two");
+
+    Internal::JITModule jit_module(module, forty_two, {});
+    
+    auto argv_wrapper = jit_module.argv_function();
+
+    Buffer<int32_t> buf(16, 16);
+    const void *args[1];
```
Comment (abadams) [inline test/correctness/external_code.cpp]: const void *args[] = {buf.raw_buffer()}; seems simpler.

Comment (abadams) [inline test/correctness/external_code.cpp]: Or given that there's one arg, you could also just grab the main function and do a function pointer cast on it to int (*)(halide_buffer_t *)

**Comment by zvookin** [comment]
I think review feedback, to date, is addressed.

File: src/ExternalCode.h [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    /** Return true if this container holds llvm bitcode linkable with
+     * code generated for the target argument. The matching is done
+     * on the architecture, bit width, and operating system
+     * only. Features are ignored. If the container is for
+     * Target::ArchUnkonwn, it applies to all architectures -- meaning
+     * it is generic llvm bitcode. If the OS is OSUnknown, it applies
+     * to all operationg systems. The bit width must match.
+     *
+     * Ignoring feature flags isn't too important since generaly
```
Comment (abadams) [inline src/ExternalCode.h]: generally

File: src/ExternalCode.h [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+     * ExternalCode will be constructed in a Generator which has
+     * access to the feature flags in effect and can select code
+     * qppropriately. */
```
Comment (abadams) [inline src/ExternalCode.h]: appropriately

**Comment by abadams** [comment]
Could probably use a generator test to act as an example, but LGTM if you want to merge it without.

**Comment by abadams** [comment]
There seem to be two issues that remain.

Target triple for the extern code blob is messed up on mingw in the jit test:

https://buildbot.halide-lang.org/master/builders/mingw64-64-trunk/builds/2878/steps/make%20test_correctness_2/logs/stdio

Not sure how that would be happening - I thought the extern code blobs were only ever added to existing Halide modules, so they should inherit the target triple.

32-bit generator test returning wrong output:

https://buildbot.halide-lang.org/master/builders/linux-32-gcc49-381/builds/284/steps/make%20test_generators/logs/stdio

**Comment by steven-johnson** [comment]
When building on Mac, we get warnings:
```
warning: Linking two modules of different data layouts: 'org.halide-lang.extern_code_extern' is 'e-m:e-v128:32-v16:16-v32:32-v96:32-n8:16:32:64-S128' whereas 'external_code' is 'e-m:o-i64:64-f80:128-n8:16:32:64-S128'

warning: Linking two modules of different target triples: org.halide-lang.extern_code_extern' is 'le64-unknown-unknown-unknown' whereas 'external_code' is 'x86_64-apple-macosx'
```

If these aren't important, we should find a way to suppress them.

**Comment by steven-johnson** [comment]
Also, multitarget output needs some love to avoid getting duplicate extern symbols. I'll work on it.

File: src/Module.cpp [hunk 10]
```diff
@@ -266,8 +275,27 @@ Module Module::resolve_submodules() const {
     for (const auto &buf : buffers()) {
         lowered_module.append(buf);
     }
+    for (const auto &ec : external_code()) {
+        lowered_module.append(ec);
+    }
     for (const auto &m : submodules()) {
         Module copy(m.resolve_submodules());
```
Comment (steven-johnson) [inline src/Module.cpp]: We make a copy with submodules resolved, and laboriously add the external code blocks to it... but then don't do anything with the copy (we compile "m", not "copy"). Is this right? If so, it needs a comment about what it accomplishes.

**Pull request merged**